apps/emqx/integration_test/emqx_persistent_session_ds_SUITE.erl

```diff
@@ -628,7 +689,124 @@ t_session_gc(Config) ->
+        }),
+        Client = start_client(Params),
+        {ok, _} = emqtt:connect(Client),
+        Client
+    end,
+
+    ?check_trace(
+        begin
+            Topic = <<"some/topic">>,
+
+            ?force_ordering(
+                #{?snk_kind := release_gc},
+                #{?snk_kind := ds_message_gc, ?snk_span := start}
+            ),
+
+            %% 1.1 Client connects, subscribes and keeps consuming msgs.
+            %% 1.2 GC runs.
+            %% 1.3 New messages are produced, client pulls again.
+            ClientId1 = <<"message_gc1">>,
+            Client1 = StartClient(ClientId1, Port1),
+
+            %% 2.1 Client connects, subscribes.
+            %% 2.2 Client consumes 1 msg, disconnects.
+            %% 2.3 GC runs and creates a gap.
+            %% 2.4 New messages are produced
+            %% 2.5 Client reconnects and pulls again.
+            ClientId2 = <<"message_gc2">>,
+            Client2 = StartClient(ClientId2, Port2),
+
+            %% 3.1 Client connects and subscribes, disconnects.
+            %% 3.2 GC runs.
+            %% 3.3 Client reconnects and starts consuming and never sees old messages.
+            ClientId3 = <<"message_gc3">>,
+            Client3 = StartClient(ClientId3, Port3),
+
+            ?tp(notice, "clients", #{c1 => Client1, c2 => Client2, c3 => Client3}),
+
+            ?tp(notice, "1.1, 2.1, 3.1", #{}),
+            lists:foreach(
+                fun(Client) ->
+                    {ok, _, [?RC_GRANTED_QOS_2]} = emqtt:subscribe(Client, Topic, ?QOS_2),
+                    ok
+                end,
+                [Client1, Client2, Client3]
+            ),
+            ?tp(notice, "3.1", #{}),
+            ok = emqtt:stop(Client3),
+            ?tp(notice, "2.2", #{}),
+            {ok, _} = emqtt:publish(Client1, Topic, <<"1">>, ?QOS_1),
+            ?assertReceive({publish, #{payload := <<"1">>, client_pid := Client2}}, 2_000),
+            ok = emqtt:stop(Client2),
+
+            ?tp(notice, "Messages for 1.1 (should be seen); 2.3, 3.3 (shouldn't)", #{}),
+            {ok, _} = emqtt:publish(Client1, Topic, <<"2">>, ?QOS_1),
+            {ok, _} = emqtt:publish(Client1, Topic, <<"3">>, ?QOS_2),
+
+            ?tp(notice, "1.2, 2.3, 3.2 (GC)", #{}),
+            ?assertWaitEvent(
+                ?tp(release_gc, #{}),
+                #{?snk_kind := message_gc_generation_dropped},
+                MessageRetentionPeriod * 3
+            ),
+
+            ?tp(notice, "Messages for 1.1, 2.4, 3.3 (all clients should see)", #{}),
+            {ok, _} = emqtt:publish(Client1, Topic, <<"4">>, ?QOS_1),
+            {ok, _} = emqtt:publish(Client1, Topic, <<"5">>, ?QOS_2),
+
+            ?tp(notice, "2.5, 3.3", #{}),
+            Client2b = StartClient(ClientId2, Port2),
+            Client3b = StartClient(ClientId3, Port3),
+            ?tp(notice, "new clients", #{c2b => Client2b, c3b => Client3b}),
+
+            ?tp(notice, "Finish", #{}),
+            ExpectedMessages =
```

Comment:
I wanted to register these cases to make it clear what the behavior will be.  Otherwise the question might be indefinite.  

Besides, doing this simple integration test uncovered problems which led to `emqx_ds_storage_layer:generation_get_safe` that otherwise wouldn't be so obvious if there weren't clients in these conditions.